### PR TITLE
openvswitch: add openssl-1.1 compatibility

### PR DESCRIPTION
**Pull request by cotequeiroz**
Do not call library initialization when compiling with openssl 1.1.

The package generates the C source files for its DH parameters at compile time using the host installed openssl.  This patch adds a DH source, using the same parameters, compatible with openssl 1.0 and 1.1.

Signed-off-by: Eneas U de Queiroz <cote2004-github@yahoo.com>

Maintainer: me / @yousong 
Compile tested: mipsel/brcm47xx, openwrt master

### CODE REVIEW

File: net/openvswitch/patches/0105-openssl-1.1.patch [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+--- a/lib/stream-ssl.c
++++ b/lib/stream-ssl.c
+@@ -951,8 +951,10 @@ do_ssl_init(void)
+     /* The following call is needed if we "#include <openssl/applink.c>". */
+     CRYPTO_malloc_init();
+ #endif
++#if OPENSSL_VERSION_NUMBER < 0x10100000L
+     SSL_library_init();
+     SSL_load_error_strings();
++#endif
```
Comment (yousong) [inline net/openvswitch/patches/0105-openssl-1.1.patch]: Why is this necessary and right?

File: net/openvswitch/patches/0105-openssl-1.1.patch [hunk 2]
```diff
@@ -0,0 +1,204 @@
+--- /dev/null
++++ b/lib/dhparams.c
```
Comment (yousong) [inline net/openvswitch/patches/0105-openssl-1.1.patch]: Is this change only about removing dependency on host openssl installation?

My local run of the `openssl dhparams` command cannot produce the same `dhp_2014` and other tables.  Can you share the implications?

**Comment by yousong** [comment]
Hi @cotequeiroz , please also send these patches to the upstream project to at least collect some opinions there.  Thank you.

File: net/openvswitch/patches/0105-openssl-1.1.patch (hunk 1 above)
Comment (cotequeiroz) [reply net/openvswitch/patches/0105-openssl-1.1.patch]: These calls were deprecated in openssl 1.1.0.  Library initialization is done automatically and there is no need to call them anymore.  So, we're only calling them with openssl < 1.1.  What this does is to allow compilation against openssl 1.1.0 compiled without deprecated API.

File: net/openvswitch/patches/0105-openssl-1.1.patch (hunk 2 above)
Comment (cotequeiroz) [reply net/openvswitch/patches/0105-openssl-1.1.patch]: I've generated the parameters with openssl 1.1, and patched it so it would compile with openssl 1.0.  I will double check.  I might have done something wrong, since travis can't build it.

File: net/openvswitch/patches/0105-openssl-1.1.patch (hunk 1 above)
Comment (yousong) [reply net/openvswitch/patches/0105-openssl-1.1.patch]: It's deprecated in 1.1 but I think there is also a replacement, at least for `SSL_library_init()`...  Is the load error strings call also deprecated?

>If you are using OpenSSL 1.0.2 or below, then you would use SSL_library_init. If you are using OpenSSL 1.1.0 or above, then you would use OPENSSL_init_ssl. A compatibility macro exists in ssl.h that maps SSL_library_init to OPENSSL_init_ssl, so you can continue to use SSL_library_init if desired.
> https://wiki.openssl.org/index.php/Library_Initialization

**Comment by yousong** [comment]
My major concern is that maybe they do the dhparam generation at run-time instead of using pre-generated dhparam.c for a reason, probably security-related, or just to make the process easily reproducible?  Either way, this should be clarified.

Host openssl binary should not be an issue as it will be provided by `tools/libressl`.

**Comment by cotequeiroz** [comment]
I have streamlined the file a little bit, so it become more readable.  It is almost straight from openssl 1.1 dhparam-generated file.  In fact, you can do it by running this (I have openssl 1.1 installed under `/usr/lib64/openssl-1.1`):

`(echo '#include "lib/dhparams.h"' && /usr/lib64/openssl-1.1/bin/openssl dhparam -C -in lib/dh1024.pem -noout && /usr/lib64/openssl-1.1/bin/openssl dhparam -C -in lib/dh2048.pem -noout && /usr/lib64/openssl-1.1/bin/openssl dhparam -C -in lib/dh4096.pem -noout) | sed -e 's/\(get_dh[0-9]*\)()/\1(void)/' -e 's/DH_set0_pqg/my_\0/' -e 's/^.*get_dh1024.*$/#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined (LIBRESSL_VERSION_NUMBER)\nstatic inline int my_DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)\n{\n    dh->p = p;\n    dh->g = g;\n    return 1;\n}\n#else\n#define my_DH_set0_pqg DH_set0_pqg\n#endif\n\n\0/' > lib/dhparams.c`

I have opened an RFC pull request upstream to get their feedback.

Cheers

**Comment by cotequeiroz** [comment]
Hello.  The first Deprecated API PR was merged upstream https://github.com/openvswitch/ovs/commit/bca4ff53aef16d38aeb1569edaaca6ac4feac6e8
I've separated the patches (Deprecated API in one, dhparams.c in the second), and updated to the one currently being looked at upstream.

**Comment by yousong** [comment]
@cotequeiroz , Thank you for the work.  Please also update the dhparams.c patch to the one in the patchwork, then we can have it in ;)

File: net/openvswitch/patches/0105-openssl-1.1.patch (hunk 1 above)
Comment (yousong) [reply net/openvswitch/patches/0105-openssl-1.1.patch]: Reply to my own comment above

> From OPENSSL_init_ssl doc
>
> As of version 1.1.0 OpenSSL will automatically allocate all resources that it needs so no explicit initialisation is required. Similarly it will also automatically deinitialise as required.

**Comment by cotequeiroz** [comment]
I've updated the patch.  I'm just not sure what's up with travis.  Apparently it mixed this up with other commits, but it does compile.

**Comment by yousong** [comment]
Merged.  Thank you for your time and efforts!
